### ISSUE TITLE
What is 'knn' in DotProductAttention Class?

### ISSUE THREAD

**Issue by jun297** [Echo0125/Memory-and-Anticipation-Transformer#4]
https://github.com/Echo0125/Memory-and-Anticipation-Transformer/blob/main/src/rekognition_online_action_detection/models/transformer/multihead_attention.py

DotProductAttention Class is different from the one in LSTR

class DotProductAttention(nn.Module):

    def __init__(self, dropout=0.0):
        super(DotProductAttention, self).__init__()

        self.dropout = dropout

    def forward(self, q, k, v, attn_mask=None, knn=False):
        B, N1, N2 = q.shape[0], q.shape[-2], k.shape[-2]
        attn_output_weights = torch.bmm(q, k.transpose(1, 2))
        if attn_mask is not None:
            attn_output_weights += attn_mask
        if knn:
            mask=torch.zeros(B,N1,N2,device=q.device,requires_grad=False)
            index=torch.topk(attn_output_weights,k=int(N2 * 3 // 4),dim=-1,largest=True)[1]
            mask.scatter_(-1,index,1.)
            # attn_output_weights = torch.where(mask>0,attn_output_weights,torch.full_like(attn_output_weights,-1e7))
            attn_output_weights = torch.where(mask > 0, attn_output_weights, torch.full_like(attn_output_weights, float('-inf')))

        attn_output_weights = F.softmax(attn_output_weights, dim=-1)
        attn_output_weights = F.dropout(attn_output_weights,
                                        p=self.dropout,
                                        training=self.training)
        attn_output = torch.bmm(attn_output_weights, v)
        return attn_output

What is this? 
is there any explanation in the paper?

**Comment by Echo0125** [comment]
We discussed this problem detailly in **Memory Dropping** in section 5.2. We also explore its effectiveness in the supplemental material.

**Comment by jun297** [comment]
Thank you!
